src/libraries/Common/src/System/Net/IPv4AddressHelper.Common.cs

```diff
@@ -105,32 +107,45 @@ internal static unsafe bool IsValid(char* name, int start, ref int end, bool all
         //                 / "2" %x30-34 DIGIT     ; 200-249
         //                 / "25" %x30-35          ; 250-255
         //
-        internal static unsafe bool IsValidCanonical(char* name, int start, ref int end, bool allowIPv6, bool notImplicitFile)
+        internal static bool IsValidCanonical<TChar>(ReadOnlySpan<TChar> name, out int bytesConsumed, bool allowIPv6, bool notImplicitFile)
+            where TChar : unmanaged, IBinaryInteger<TChar>
         {
+            TChar ComponentSeparator = TChar.CreateChecked('.');
+            ReadOnlySpan<TChar> PrefixSeparators = [TChar.CreateChecked('/'), TChar.CreateChecked('\\')];
+            ReadOnlySpan<TChar> UrlSeparators = [TChar.CreateChecked(':'), TChar.CreateChecked('?'), TChar.CreateChecked('#')];
+            TChar OctalPrefix = TChar.CreateChecked('0');
+
+            ReadOnlySpan<TChar> IPv6Terminators = [TChar.CreateChecked('%'), TChar.CreateChecked('/'), TChar.CreateChecked(']')];
+
             int dots = 0;
             int number = 0;
             bool haveNumber = false;
             bool firstCharIsZero = false;
+            int current;
 
-            while (start < end)
+            Debug.Assert(typeof(TChar) == typeof(char) || typeof(TChar) == typeof(byte));
+
+            bytesConsumed = 0;
+            for (current = 0; current < name.Length; current++)
             {
-                char ch = name[start];
+                TChar ch = name[current];
+
                 if (allowIPv6)
                 {
                     // for ipv4 inside ipv6 the terminator is either ScopeId, prefix or ipv6 terminator
-                    if (ch == ']' || ch == '/' || ch == '%')
+                    if (IPv6Terminators.Contains(ch))
                         break;
                 }
-                else if (ch == '/' || ch == '\\' || (notImplicitFile && (ch == ':' || ch == '?' || ch == '#')))
+                else if (PrefixSeparators.Contains(ch) || (notImplicitFile && UrlSeparators.Contains(ch)))
```

Comment:
Calling `Contains` to check for two possible values is likely going to be quite a bit slower than the existing code. This should be better inlined check against two values like before.

Reply:
(multiple places)

Reply:
You're right. I was actually trying to make it more readable and perhaps slightly faster - I'm pretty sure I've seen SSE intrinsics which can search for a small value in a 16-/32-bit space, and I thought ROS.Contains would use them - but a benchmark puts it at 10x slower per operation. I'll re-inline.